patches/server/0968-Add-FeatureFlag-API.patch

```diff
@@ -0,0 +1,138 @@
++
++import com.google.common.collect.BiMap;
++import com.google.common.collect.HashBiMap;
++import java.util.Map;
++import java.util.Set;
++import java.util.stream.Collectors;
++import net.minecraft.server.MinecraftServer;
++import net.minecraft.world.flag.FeatureElement;
++import net.minecraft.world.flag.FeatureFlagSet;
++import org.bukkit.Material;
++import org.bukkit.craftbukkit.util.CraftMagicNumbers;
++import org.bukkit.entity.EntityType;
++
++public enum PaperFeatureFlagProviderImpl implements FeatureFlags.FeatureFlagProvider {
++    INSTANCE;
++
++    PaperFeatureFlagProviderImpl() {
++        FeatureFlags.INSTANCE = this;
++    }
++
++    static final BiMap<FeatureFlag, net.minecraft.world.flag.FeatureFlag> FLAGS = HashBiMap.create(Map.of(
++        FeatureFlags.VANILLA, net.minecraft.world.flag.FeatureFlags.VANILLA,
++        FeatureFlags.BUNDLE, net.minecraft.world.flag.FeatureFlags.BUNDLE,
++        FeatureFlags.UPDATE_1_20, net.minecraft.world.flag.FeatureFlags.UPDATE_1_20
++    ));
++
++    @Override
++    public boolean isEnabled(final FeatureDependant dependant) {
++        final FeatureFlagSet set = MinecraftServer.getServer().getWorldData().enabledFeatures();
++        return getFeatureElement(dependant).isEnabled(set);
++    }
++
++    @Override
++    public Set<FeatureFlag> requiredFeatures(final FeatureDependant dependant) {
++        final FeatureFlagSet requiredFeatures = getFeatureElement(dependant).requiredFeatures();
++        return FLAGS.entrySet().stream().filter(entry -> requiredFeatures.contains(entry.getValue())).map(Map.Entry::getKey).collect(Collectors.toUnmodifiableSet());
++    }
++
++    static FeatureElement getFeatureElement(final FeatureDependant dependant) {
++        if (dependant instanceof Material material) {
++            if (material.isItem()) {
++                return CraftMagicNumbers.getItem(material);
++            } else if (material.isBlock()) {
++                return CraftMagicNumbers.getBlock(material);
++            } else {
++                throw new IllegalArgumentException(material + " must be a block or item");
++            }
++        } else if (dependant instanceof EntityType entityType) {
++            return CraftMagicNumbers.getEntityTypes(entityType);
++        } else {
++            throw new IllegalArgumentException(dependant + " is not a valid feature dependant");
++        }
++    }
++}
+diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
+index 781b72fd88149642c9fceaecfbfe7546273fb749..25726d007d8f0875bf86faa835717017edda0415 100644
+--- a/src/main/java/net/minecraft/server/Main.java
++++ b/src/main/java/net/minecraft/server/Main.java
+@@ -111,6 +111,7 @@ public class Main {
+                 JvmProfiler.INSTANCE.start(Environment.SERVER);
+             }
+ 
++            io.papermc.paper.world.flag.PaperFeatureFlagProviderImpl.INSTANCE.getClass(); // Paper - init FeatureFlagProvider
```

Comment:
Is this the preferred way we want to inject server singletons into the API ?
I know this is a somewhat long standing issue between services, the server interface and this one.